src/blockdev.rs

```diff
@@ -492,38 +492,65 @@ impl Drop for Mount {
 
 #[derive(Debug)]
 pub struct SavedPartitions {
-    sector_size: Option<u64>,
+    sector_size: u64,
     partitions: Vec<(u32, GPTPartitionEntry)>,
 }
 
 impl SavedPartitions {
     pub fn new(disk: &mut File, filters: &[PartitionFilter]) -> Result<Self> {
-        let mut result = Self {
-            sector_size: None,
-            partitions: Vec::new(),
+        // try reading GPT
+        let maybe_gpt = match GPT::find_from(disk) {
+            Ok(gpt) => Some(gpt),
+            // no GPT on this disk, so no partitions to save
+            Err(gptman::Error::InvalidSignature) => None,
+            Err(e) => return Err(e).chain_err(|| "reading partition table")?,
         };
 
-        // skip reading partition table if we don't have an accept filter
-        if filters.is_empty() {
-            return Ok(result);
-        }
+        // determine sector size
+        let sector_size = if disk
+            .metadata()
+            .chain_err(|| "getting disk metadata")?
+            .file_type()
+            .is_block_device()
+        {
+            // We have a real disk, so treat it as the source of truth.
+            get_sector_size(&disk)?.get() as u64
+        } else if let Some(ref gpt) = maybe_gpt {
+            // No real disk, but an existing GPT.  Trust it.
+            gpt.sector_size
+        } else {
+            // We're not reading from a real disk, and there's no existing
+            // GPT.  This should only happen in unit tests.  Pick a default.
+            512
+        };
 
-        // read GPT
-        let gpt = match GPT::find_from(disk) {
-            Ok(gpt) => gpt,
-            // no GPT on this disk, so no partitions to save
-            Err(gptman::Error::InvalidSignature) => return Ok(result),
-            Err(e) => return Err(e).chain_err(|| "reading partition table")?,
+        // early return if no GPT
+        let gpt = match maybe_gpt {
+            Some(gpt) => gpt,
+            None => {
+                return Ok(Self {
+                    sector_size,
+                    partitions: Vec::new(),
+                })
+            }
         };
-        Self::verify_disk_sector_size_matches_gpt(disk, gpt.sector_size)?;
-        result.sector_size = Some(gpt.sector_size);
+
+        // cross-check disk and GPT sector sizes if we have both
+        Self::verify_disk_sector_size(disk, gpt.sector_size)?;
 
         // save partitions accepted by filters
-        for (i, p) in gpt.iter() {
-            if Self::matches_filters(i, p, filters) {
-                result.partitions.push((i, p.clone()));
+        let mut partitions = Vec::new();
+        if !filters.is_empty() {
+            for (i, p) in gpt.iter() {
```

Comment:
If `gpt` is not used afterward, an `into_iter()` here may work to remove the need of a `clone` below.

Reply:
`GPT` doesn't implement `into_iter()`.